src/coreclr/jit/valuenum.h

```diff
@@ -529,6 +529,114 @@ class ValueNumStore
 
     void PeelOffsets(ValueNum* vn, target_ssize_t* offset);
 
+    typedef JitHashTable<ValueNum, JitSmallPrimitiveKeyFuncs<ValueNum>, bool> ValueNumSet;
+
+    class SmallValueNumSet
+    {
+        union
+        {
+            ValueNum     m_inlineElements[4];
+            ValueNumSet* m_set;
+        };
+        unsigned m_numElements = 0;
+
+    public:
+        unsigned Count()
+        {
+            return m_numElements;
+        }
+
+        template <typename Func>
+        void ForEach(Func func)
+        {
+            if (m_numElements <= ArrLen(m_inlineElements))
+            {
+                for (unsigned i = 0; i < m_numElements; i++)
+                {
+                    func(m_inlineElements[i]);
+                }
+            }
+            else
+            {
+                for (ValueNum vn : ValueNumSet::KeyIteration(m_set))
+                {
+                    func(vn);
+                }
+            }
+        }
+
+        // Returns false if the value wasn't found
+        bool Lookup(ValueNum vn);
+
+        // Returns false if the value already exists
+        bool Add(Compiler* comp, ValueNum vn);
+    };
+
+    enum class VNVisit
+    {
+        Continue,
+        Abort,
+    };
+
+    enum class VNVisitResult
+    {
+        Completed,
+        Aborted,
+        MaxDepthReached,
+    };
```

Comment:
`MaxDepthReached` is unused now so looks like this enum can be removed

Reply:
I think it's better to keep the separate enum because `Completed` sounds better than `Continue` + ability to extend it in future

Reply:
or, probably, at this point we can convert it back to bool ?

Reply:
Well, IMO it's better to be consistent with the other visitors, e.g. `GenTreeVisitor::WalkTree` and `BasicBlockVisit` etc. just return the same type of visit result. With that said I'm fine either way.

Reply:
Ok, makes sense to have a consistency with WalkTree then